### PR TITLE
Added dynamically updating atomgroup selections (#175 and #1074)

### LINKED ISSUE #175: Interchangeability of AtomGroup and selection strings
Following the discussion on Issue #171 it would be interesting to adapt any functions that take AtomGroups to also be able to take selection strings (and vice versa).

I brought it up because an analysis class I use often, `LeafletFinder`, accepts exclusively selection strings. I have a fix for this particular case, and since I'm working on Issue #174 I'll look into allowing `AtomGroups` to be passed to selection functions.

Original issue reported on code.google.com by `manuel.n...@gmail.com` on 28 Mar 2014 at 1:16

### PR DESCRIPTION
**Pull request by mnmelo**
Doesn't really close #175 and #1074, but that's where the idea for this feature matured.

It's been a while since I had to check documentation, and I wasn't able to build the sphinx html locally for testing. The [wiki instructions](https://github.com/MDAnalysis/mdanalysis/wiki/WritingDocumentation#build-the-docs-with-sphinx) didn't really work. Any hints?

Also:
- Cleaned up some selection docs
- Updated some uses of itertools.izip to six.zip
- Cleaned up some variable names in groups.py that could eventually clash with
the six module.

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

**Comment by mnmelo** [comment]
So, something seems a bit off, in that the code didn't seem to be touched by the tests I wrote. I'll double check.

**Comment by mnmelo** [comment]
Ok. Silly mistake. This should do it.

File: package/MDAnalysis/core/groups.py [hunk 1]
```diff
@@ -1876,6 +1932,65 @@ def __getattr__(self, attr):
...
+    def update_selection(self):
+        bg = self._base_group
+        sels = self._updating_selections
+        ix = sum([sel.apply(bg) for sel in sels])._ix
+        super(UpdatingAtomGroup, self).__init__(ix, bg._u)
+        self._lastupdate = bg._u.trajectory.ts.frame
+
+    def __getattribute__(self,name):
+        if not name in ("_base_group","_updating_selections","_lastupdate"):
+            try:
```
Comment (richardjgowers) [inline package/MDAnalysis/core/groups.py]: Universe will remain constant across updates, so this could probably get set at init time.  With `_last_update` we could set that to an impossible value (-1?) on init so that the equality check always fails first time round.  Then this try/except block doesn't have to exist.

File: package/MDAnalysis/core/groups.py [hunk 2]
```diff
@@ -170,7 +172,10 @@ def _mix(cls, other):
             A class of parents :class:`_ImmutableBase`, *other* and this class.
             Its name is the same as *other*'s.
         """
-        return type(other.__name__, (_ImmutableBase, other, cls), {})
+        if '__new__' in other.__dict__:
```
Comment (richardjgowers) [inline package/MDAnalysis/core/groups.py]: You'll have to put in a note about which classes will follow route here

**Comment by mnmelo** [comment]
And any tips on how to view the docs?

File: package/MDAnalysis/core/groups.py (hunk 2 above)
Comment (mnmelo) [inline package/MDAnalysis/core/groups.py]: Ok. I added extra text to the docstring, but agree it's not as explicit.

**Comment by richardjgowers** [comment]
For docs, I've always followed the instructions in the wiki ok, what's the error you're getting?

This error means you need to install sphinx
```
sphinx-build -b html   source ../html
make: sphinx-build: Command not found
make: *** [html] Error 127
```

This error means that sphinx can't see the cython modules (ie do a developer install maybe)
```
Exception occurred:
  File "/home/richard/miniconda2/envs/mdadev/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
ImportError: No module named c_distances
```

**Comment by mnmelo** [comment]
Yeah, sorry for being terse. It's the second one. Thanks!

**Comment by tylerjereddy** [comment]
Does this enhancement improve performance or is it primarily designed for syntactic convenience? For example, is a dynamic selection faster than a regenerated conventional selection as you iterate through a trajectory? Conversely, if you use a dynamic selection when a conventional selection would have sufficed, is there a performance penalty for this during trajectory iteration?

Will the dynamic selection initialization machinery recognize when it is an atomgroup / selection that is immutable (so dynamic updates aren't needed)?

**Comment by mnmelo** [comment]
There is a (probably small) performance improvement because `UpdatingAtomGroup` stores the `Selection` objects, and therefore you don't need to build the parse tree every time.

There's no check if the selection might be static. Can you think of a nifty way to do it?

**Comment by richardjgowers** [comment]
I imagine this will be mostly syntactic, these will mostly be used for distance based selections I guess, and it's the `distance_array`/`KDTree` calls that will be slow.  That said, I do like the idea of a "select water around this protein" object in MDA.

**Comment by tylerjereddy** [comment]
I suspect it would be difficult to check if a selection is static, although you could probably establish a set of rules to identify a selection that is guaranteed to be static, while for others you'd have to be dynamic just to be safe. 

For example, "around" could be static if the selection of waters around a protein never actually changes during a given trajectory, but it would not be a safe assumption and you'd have to iterate the entire trajectory to verify.

Conversely, I suspect that a subset of selections such as "resid 25" are largely static / immutable, so one could check for the exclusion of all possible dynamic-enabling keywords in selection strings. I wonder if it is really worth it though. The other thing is that apparently the new GROMACS TNG trajectory format ( #865 ) supports topology mutations during a trajectory (grand canonical simulations of some sort?), so to be completely forward thinking it may be a bad idea to try and make assumptions about which selections are static or dynamic.

**Comment by richardjgowers** [comment]
If someone's requested a dynamic selection via a kwarg then we should treat that literally.  They could for example be using a random attribute (resid?) to label atoms inside their loop, which they want the dynamic selection to respect.

tl;dr the performance boost with guessing static things seem small compared to the possible headaches :D

**Comment by mnmelo** [comment]
I agree with @richardjgowers. Also, stuff may look like a static selection (say `"resnum < 10"`) but if the base group is an `UpdatingAtomGroup` the outcome will still be dynamic.

Performance-wise, I forgot to mention that a very small hit can be expected because of the `__getattribute__` layer that has to be put in place. It's just a couple calls per attribute accession, and unlikely to be noticeable unless accessing extremely frequently.

**Comment by mnmelo** [comment]
Ok, take a look if this seems good. In the end I found a way to simplify the `UpdatingAtomGroup` argument signature to just an `AtomGroup` and a tuple of `Selections`.

For reference, my problems with sphinx were due to python3 sphinx version conflicts. I'll add that to the instructions page.

File: package/MDAnalysis/core/groups.py [hunk 3]
```diff
@@ -1876,6 +1954,67 @@ def __getattr__(self, attr):
...
+        self._base_group = base_group
+        self._updating_selections = selections
+        self._lastupdate = None
+        self.update_selection()
+
+    def update_selection(self):
+        bg = self._base_group
+        sels = self._updating_selections
+        ix = sum([sel.apply(bg) for sel in sels])._ix
+        super(UpdatingAtomGroup, self).__init__(ix, bg._u)
```
Comment (richardjgowers) [inline package/MDAnalysis/core/groups.py]: If we just changed the `._ix` attribute in place, would that clean up a lot of the `__new__` magic that is required elsewhere?  I'm not sure we really need to do the `super` call.

File: package/CHANGELOG [hunk 4]
```diff
@@ -14,11 +14,13 @@ The rules for this file:
...
 ------------------------------------------------------------------------------
 ??/??/16 kain88-de, fiona-naughton, richardjgowers, tyler.je.reddy, jdetle
-         euhruska, orbeckst, rbrtdlg, jbarnoud, wouterboomsma, shanmbic
+         euhruska, orbeckst, rbrtdlg, jbarnoud, wouterboomsma, shanmbic,
+         manuel.nuno.melo
 
   * 0.16.0
 
 Enhancements
+  * Added dynamic selections (addreses Issues #175 and #1074).
```
Comment (orbeckst) [inline package/CHANGELOG]: addresses

File: package/MDAnalysis/core/groups.py [hunk 5]
```diff
@@ -155,9 +157,11 @@ def _subclass(cls, singular=None):
     def _mix(cls, other):
         """Creates a subclass with ourselves and another class as parents.
 
-        Classes mixed at this point always override :meth:`__new__`, causing
-        further instantiations to shortcut to :meth:`object.__new__` (skipping
-        the cache-fetch process for :class:`_MutableBase` subclasses).
+        Classes mixed at this point override :meth:`__new__`, causing further
+        instantiations to shortcut to :meth:`object.__new__` (skipping the
+        cache-fetch process for :class:`_MutableBase` subclasses). An
+        exception is if the *other* class itself already has a __new__ method,
```
Comment (orbeckst) [inline package/MDAnalysis/core/groups.py]:
```reST
:meth:`~object.__new__`
```
instead of `__new__` which formats as __new__ (bold)

File: package/MDAnalysis/core/groups.py [hunk 6]
```diff
@@ -1033,6 +1062,9 @@ def select_atoms(self, sel, *othersel, **selgroups):
 
         Subselections can be grouped with parentheses.
 
+        Selections can be set to update automatically on frame change, by
+        setting the *updating* named argument to `True`.
```
Comment (orbeckst) [inline package/MDAnalysis/core/groups.py]:
```reST
`updating`
```

File: package/MDAnalysis/core/groups.py [hunk 7]
```diff
@@ -1182,17 +1214,63 @@ def select_atoms(self, sel, *othersel, **selgroups):
                 therefore have atoms that were initially absent from the
                 instance :meth:`~select_atoms` was called from.
 
+            global *selection*
```
Comment (orbeckst) [inline package/MDAnalysis/core/groups.py]: With numpy docs, all arguments that are mentioned in the Arguments list can be written in single backtics.

File: package/MDAnalysis/core/groups.py [hunk 8]
```diff
@@ -1182,17 +1214,63 @@ def select_atoms(self, sel, *othersel, **selgroups):
...
+           ``global group``.
         .. versionchanged:: 0.13.0
            Added *bonded* selection
         .. versionchanged:: 0.16.0
            Resid selection now takes icodes into account where present.
+        .. versionadded:: 0.16.0
+           Updating selections now possible by setting the ``updating`` argument.
 
         """
+        updating = selgroups.pop('updating',False)
```
Comment (orbeckst) [inline package/MDAnalysis/core/groups.py]: PEP8: space after comma

File: package/MDAnalysis/core/groups.py [hunk 9]
```diff
@@ -1876,6 +1954,67 @@ def __getattr__(self, attr):
...
+
+    def update_selection(self):
+        bg = self._base_group
+        sels = self._updating_selections
+        ix = sum([sel.apply(bg) for sel in sels])._ix
+        super(UpdatingAtomGroup, self).__init__(ix, bg._u)
+        self._lastupdate = bg._u.trajectory.ts.frame
+
+    def __getattribute__(self,name):
+        if not name in ("_base_group","_updating_selections","_lastupdate"):
```
Comment (orbeckst) [inline package/MDAnalysis/core/groups.py]: PEP8 comma space

File: package/MDAnalysis/core/groups.py [hunk 10]
```diff
@@ -1876,6 +1954,67 @@ def __getattr__(self, attr):
...
+            if (object.__getattribute__(self, "_u").trajectory.frame !=
```
Comment (orbeckst) [inline package/MDAnalysis/core/groups.py]: Do you like `object.__gettattribute__()` better than `super(UpdatingAtomGroup, self).__getattribute__()`?

File: package/doc/sphinx/source/documentation_pages/selections.rst [hunk 11]
```diff
@@ -235,6 +237,56 @@ global *selection*
     :class:`~MDAnalysis.core.universe.Universe`, ``global`` is ignored.
```
Comment (orbeckst) [inline package/doc/sphinx/source/documentation_pages/selections.rst]: Thanks for the docs!!

File: package/MDAnalysis/core/groups.py (hunk 3 above)
Comment (mnmelo) [inline package/MDAnalysis/core/groups.py]: I'd say yes. I wanted to keep the mechanism independent of the AtomGroup/GroupBase internals, but coupling both is reasonable and indeed should clarify the code.

I'll give it a try.

File: package/MDAnalysis/core/groups.py (hunk 10 above)
Comment (mnmelo) [inline package/MDAnalysis/core/groups.py]: I'm not sure here. The `super` alternative might expose us to problems if a `__getattribute__` is ever introduced in the hierarchy chain. On the other hand, if we ever do add `__getattribute__` to `AtomGroup` (which we shouldn't because of performance) we might want it to also be triggered.

I'll try your way, hopeful that any problems that might arise will be easy to deal with.

Comment (mnmelo) [inline package/MDAnalysis/core/groups.py]: Actually, scratch what I wrote: let's stick to `object.__gettattribute__()` for easier readability and probably faster access. If a need for `super` comes up in the future, it's easier to change.

File: package/MDAnalysis/core/groups.py (hunk 3 above)
Comment (mnmelo) [inline package/MDAnalysis/core/groups.py]: Ok, no it won't do.
We need the `__new__` for stuff like slicing operations, which internally trigger a call to
`self.__class__(ix, u)`. The `__new__` is needed to divert these cases to a different class altogether (`AtomGroup`).

As for the `super` call, it can be done away with. The only problem might be future changes to `GroupBase.__init__`, that might make its initialization less simple. I already notice that one should also set `self._cache` to the empty dict, and this should be copied into `UpdatingAtomGroup` too. `super` avoids these duplications.
Any other suggestions on how to simplify?

File: package/MDAnalysis/core/groups.py [hunk 12]
```diff
@@ -1876,6 +1954,67 @@ def __getattr__(self, attr):
...
+        base_group : :class:`AtomGroup`
+            group on which *selections* are to be applied.
+        selections : a tuple of :class:`~MDAnalysis.core.selection.Selection` instances
+            selections ready to be applied to *base_group*.
+
+        """
+        # Because we're implementing __getattribute__, which needs _u for
+        # its check, no self.attribute access can be made before this line
+        self._u = base_group._u
+        self._base_group = base_group
```
Comment (richardjgowers) [inline package/MDAnalysis/core/groups.py]: Thinking forwards about pickling, if an `UpdatingGroup` just pickled the base group with itself, then even if `base_group` were also dynamic it would just recursively pickle until it hit a real `AtomGroup`.

File: package/MDAnalysis/core/groups.py (hunk 7 above)
Comment (orbeckst) [inline package/MDAnalysis/core/groups.py]: I might have misread this one: here *selection* is not the kw but the content place holder, isn't it? I can't remember if there's anything specific prescribed so then leave it as we had it. I thought that this one here was similar to https://github.com/MDAnalysis/mdanalysis/pull/1080#discussion_r89420562 ; in any case, that's only docs.

Comment (mnmelo) [inline package/MDAnalysis/core/groups.py]: Here _selection_ is indeed the kwarg name, and not a selection expression/keyword. I'll put it in backticks for correctness but am not sure that having it formatted as such in the docs helps much, since the user has nowhere to know what that kwarg is (it's only explained in the doc  body of the `global` keyword).

Comment (mnmelo) [inline package/MDAnalysis/core/groups.py]: Sorry, my previous comment was totally off (mistook the `global` keyword for `group`). As you said, `selection` is just the placeholder. I'll leave it as `*selection*`.

Incidentally, the `group *group-name*` docs should be the ones changed to backticks, since `group-name` refers to a kwarg name, right?

**Comment by orbeckst** [comment]
Yes, although it's a meta identifier, isn't it? Anyway, try back tics and if it looks awful we change it again. It's really the most minor of minor comments that I have ;-)

--
Oliver Beckstein
email: orbeckst@gmail.com


> Am Nov 28, 2016 um 9:27 schrieb mnmelo <notifications@github.com>:
> 
> Incidentally, the group *group-name* docs should be the ones changed to backticks, since group-name refers to a kwarg name, right?

**Comment by mnmelo** [comment]
Ok, I was able to adapt the `UpdatingAtomGroup` class so that it no longer needs its own `__new__`.
In addition, and more importantly, I added the possibility for regular `AtomGroups` to be made updating simply by doing:
```python
ag = u.select_atoms("prop z > 50")
ag.updating = True
``` 
For this to work I made all `AtomGroups` stemming from `select_atoms` store their base groups and selection objects. I believe these are all relatively shallow and cheap references to keep, but want your opinion on it.

**Comment by richardjgowers** [comment]
I need to have a real play with this this weekend (and see if I can break it :D).  My initial thoughts are that if you want to start including extra attributes in an AtomGroup, we should have an organised place to store them.  Otherwise copying/pickling/slicing? AtomGroups gets much more difficult.

That said, I can't really think of a situation where you would create an AtomGroup and then want to make it updating after that?  I think the kwarg for `select_atoms` should be enough?

Anyway, thanks @mnmelo I'll review this properly at the weekend

**Comment by mnmelo** [comment]
Thanks for the time!

For me the use case for being able to set updating after `select_atoms` is that an analysis might require an updating AtomGroup, and set it itself, and the user needn't worry about it.

OTOH, it does introduce some more complexity even into the standard AtomGroup, and I'll be happy to leave it as the `select_atoms` kwarg. (It also requires a cool but relatively hackish class change, which wet might want to avoid.)

Finally, I got around the need for `UpdatingAtomGroup.__new__` by having each instance store a pointer to the class that should be used for slicing/addition results. Let me know if you find this a better solution.

**Comment by mnmelo** [comment]
@richardjgowers, I'm setting the milestone on this for 0.16. I'm not in a particular hurry to get this out with 0.16, but I think that it makes sense then as it is a nice feature addition. It also got recent attention from a similar feature request (#1118), so there's definitely a user base for it.

In any case, please just undo the milestone assignment if you disagree.

**Comment by richardjgowers** [comment]
@mnmelo yeah it looks ready, it's just waiting on a lazy reviewer :D

**Comment by mnmelo** [comment]
No worries! And besides, I really want your non-rushed feedback on the hacking I now have in place, and whether I should roll back to a more conservative approach.

**Comment by dotsdl** [comment]
Reviewing this as well. I'm really paranoid about introducing additional state into Groups, and haven't given this a proper look yet. Trying to set aside some time to do it. :/

**Comment by dotsdl** [comment]
I did a quick play with this feature, and here's my [hot take in gist form](https://gist.github.com/dotsdl/b93f9150d8d57004165e5ae899f66dbd).

The (even shorter) gist: I don't think this is a route we should go down, because:
1. It introduces two types of `AtomGroup`s, which are not always interchangeable in terms of switching from one to another. I think this amounts to some interface shear, adding complexity that doesn't always work as one might hope.
2. It introduces more state into an `AtomGroup`, making them (slightly?) heavier.
3. It could make parallelization even more of a headache, since `AtomGroup`s now need to know which `AtomGroup` they came from to behave consistently with this feature.
4. There isn't a demonstrated performance gain, and syntactically there's not much of one, either (in my opinion). The syntactic gain isn't outweighed by the added complexity.

**Comment by mnmelo** [comment]
I feared that indeed I was piling too much on `AtomGroup`. But you mean ditching the `UpdatingAtomGroup` altogether, or just my latest iteration? (The one that adds more state to `AtomGroup` so that `ag.updating = True` works to make an AtomGroup dynamic post-facto.)

Because if I just ditch that last part (so that an `UpdatingAtomGroup` can only be made by `select_atoms`) `AtomGroups` will no longer need to hold pointers to `Selections` or any other `AtomGroups`. In fact, if I do it like that the only thing that needs to be added outside `UpdatingAtomGroup` is a class pointer to itself (`_derived_class`), so that we can customize which class comes out of a slicing/addition operation and can obviate the use of `UpdatingAtomGroup.__new__`.

Under that alternative I'd say the syntactic benefit of having the `select_atoms(selection, updating=True)` functionality  is definitely worth it.

**Comment by orbeckst** [comment]
> • It introduces more state into an AtomGroup, making them (slightly?) heavier.

It might be helpful if the philosophy of the new topology system was laid out clearly. What you are saying implies that "light weight, no state" was a primary design goal. None of this has really been fully communicated and it would be important to help other developers to navigate the new system.

I know that @richardjgowers  and @dotsdl  have done an awful lot of work on #363 but in order to keep development orderly it is high time that there are proper docs that help the other devs as well to navigate the code, i.e., #1078 . It will become unproductive and annoying really soon if the devs are not all on the same page.

>	• There isn't a demonstrated performance gain, and syntactically there's not much of one, either (in my opinion). The syntactic gain isn't outweighed by the added complexity.

IIRC one idea was that we (might) want to be able to be able to provide AtomGroups wherever we currently pass selection strings around #1074 . In this case it seemed convenient to have AGs that know that they should update. In particular, there is no way to get the same effect with a kwarg `update=True` and passing a standard static AG. But that is really under discussion.

All that said, this discussion is important and we need to figure out what our priorities are. I recognize that this can come with a lot of work for trying out many different approaches. But I don't see a way around this: prototyping and testing and discussing is what you need to do in order to make a good product.

**Comment by dotsdl** [comment]
@mnmelo yeah, I think at least making `UpdatingAtomGroup` immutable in the sense that it doesn't have the `updating` property is a step I'm in favor of, and same for `AtomGroup` itself to keep it light and relatively stateless as it already is. I'm still a bit worried about serialization, since in the current implementation the `UpdatingAtomGroup` keeps track of where it came from, and likewise up the chain. This is cool, but not sure how complicated this makes #878 more difficult to solve sanely.

**Comment by dotsdl** [comment]
@orbeckst, agreed. Moving this to top of priorities list. It's clear that this can help us to clear the air on what direction we are moving in together, which is probably a better use of time for me right now than writing new code. #technicaldebt

File: package/MDAnalysis/coordinates/memory.py [hunk 13]
```diff
@@ -352,8 +352,11 @@ def timeseries(self, asel=None, start=0, stop=-1, step=1, format='afc'):
         # Return a view if either:
         #   1) asel is None
         #   2) asel corresponds to the selection of all atoms.
+        # The 2nd case may need a deep comparison because the "all" selection
+        # creates a new AtomGroup object that isn't the same as universe.atoms.
         array = array[basic_slice]
-        if (asel is None or asel is asel.universe.atoms):
+        if (asel is None or asel is asel.universe.atoms or
+                np.array_equal(asel.indices, asel.universe.atoms.indices)):
```
Comment (richardjgowers) [inline package/MDAnalysis/coordinates/memory.py]: The `array_equal` check here will scale with system size as it will check (potentially) all array elements.  Is it necessary to include it?

File: package/MDAnalysis/core/groups.py [hunk 14]
```diff
@@ -1923,6 +1996,128 @@ def __getattr__(self, attr):
...
+        return status
+
+    def __getattribute__(self, name):
+        if not (name in object.__getattribute__(self, "_shortcut_attrs")):
+            self.ensure_updated()
+        return object.__getattribute__(self, name)
+
+    def __repr__(self):
+        name = self.level.name
+        return ("<UpdatingAtomGroup with {} {}s>"
```
Comment (richardjgowers) [inline package/MDAnalysis/core/groups.py]: Maybe include the selection string as part of the repr?  `<UpdatingAtomGroup for 'around 5.0 protein', currently with {} atoms>`

Comment (mnmelo) [inline package/MDAnalysis/core/groups.py]: Good idea. Right now only the parsed selections make it to the `UpdatingAtomGroup` but it won't be much trouble to pass the strings through as well; it's a lightweight extra.

The meaning of `__repr__` might become a bit misleading if `_base_group != u.atoms` but I think the user can live with it (or maybe I add a note to the text).

Comment (kain88-de) [inline package/MDAnalysis/core/groups.py]: How would the  `around 5.0 protein` string work together with atomgroups derived from this one. `new_ag = updating_ag[:10]`?

Comment (mnmelo) [inline package/MDAnalysis/core/groups.py]: As it is now, if you slice it you go back to having a plain `AtomGroup`, so no fancy `__repr__` either.

File: package/MDAnalysis/coordinates/memory.py (hunk 13 above)
Comment (mnmelo) [inline package/MDAnalysis/coordinates/memory.py]: For the sake of moving the UpdatingAtomGroup forward I'll roll back this change and raise an issue with the MemoryReader.

**Comment by mnmelo** [comment]
Ok, addressed @richardjgowers comments; also made `AtomGroup.__repr__` fancier (it now respects plural vs. singular number of elements) and added further testing for the `__repr__`.

File: testsuite/MDAnalysisTests/core/test_atomgroup.py [hunk 15]
```diff
@@ -331,3 +331,38 @@ def test_center_wrong_shape(self):
         weights = np.ones((self.ag.n_atoms, 2))
 
         assert_raises(TypeError, self.ag.center, weights)
+
+class TestRepresentations(object):
+    def setUp(self):
+        self.u = mda.Universe(CONECT)
```
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/core/test_atomgroup.py]: These tests are fine, but you don't really need to use `CONECT` here, you could use `make_Universe` to make a Universe that doesn't require file IO and should be a lot faster.

**Comment by richardjgowers** [comment]
@mnmelo I tried to break this with some tests (made an `UpdatingAtomGroup` which updates from an `UpdatingAtomGroup`) and it still works, so it's looking good!  There's PR #1128 that changes a few things to make it more readable, if you agree merge, otherwise close it.

Then the `.c` files that got committed need removing from this PR before it can get merged.

**Comment by mnmelo** [comment]
Awesome work with the tests @richardjgowers!
I'l merge #1128 locally, take care of those wayward `.c` files and push back here.

**Comment by mnmelo** [comment]
Hmm... wait, pushed too soon, before fixing that CONECT test case.

**Comment by mnmelo** [comment]
There, this should do it.

File: package/MDAnalysis/core/groups.py [hunk 16]
```diff
@@ -1923,6 +2000,149 @@ def __getattr__(self, attr):
...
+        if value:
+            try:
+                self._lastupdate = self._u.trajectory.frame
+            except AttributeError: # self._u has no trajectory
+                self._lastupdate = -1
+        else:
+            # This always marks the selection as outdated
+            self._lastupdate = None
+
+    def ensure_updated(self):
```
Comment (richardjgowers) [inline package/MDAnalysis/core/groups.py]: So my final (I promise) comment is that maybe this method should get demoted to `_ensure_updated` to avoid potential confusion.  I think the code block below is perfectly reasonable, but won't work as `ensure_updated` doesn't detect when the `ts` was modified, just frame numbers.

```python
uag = u.select_atoms('something', updating=True)

for ts in u.trajectory:
    # access the UAG here for something minor
    something = uag.names

    # manipulate the trajectory, maybe a fitting or PBC transformation..
    u.atoms.pack_into_box()

    # now I want to ensure my selection is updated, so I guess I'll call..
    uag.ensure_updated()

    do_heavy_analysis(uag)  # whoops, this isn't actually updated though
```

I'd rather just give a single method for this (`update_selection`) rather than two which sound like they do the same thing, but are subtlely different.

**Comment by mnmelo** [comment]
Hey sorry, I saw the 'approved' tag and didn't realize there were conflicts to tackle. Gimme a moment.

**Comment by kain88-de** [comment]
@mnmelo for the future please rebase instead of using a merge to update against develop. rebasing makes the history much more readable.

**Comment by mnmelo** [comment]
I thought as much, but
1. GitHub makes direct editing so much more convenient (and that gets added as a merge);
2. With the "Squash and Merge" option the history still stays relatively clean, no?

**Comment by mnmelo** [comment]
Just a question (maybe I missed the relevant memo): now that we are requesting changes to be pre-approved do I still need to wait for one of the assignees to merge into develop or can I go ahead and do it myself?

**Comment by kain88-de** [comment]
@mnmelo you should be able to do it yourself now since the PR has an apprved review.

File: package/MDAnalysis/core/groups.py (hunk 16 above)
Comment (richardjgowers) [inline package/MDAnalysis/core/groups.py]: @mnmelo did you think it is worth doing this or am I being too safe?

Comment (mnmelo) [inline package/MDAnalysis/core/groups.py]: Ah, yes, makes perfect sense and I overlooked your comment. I'll add it (now as another PR) ASAP.
